src/mame/barcrest/mpu1.cpp

```diff
@@ -0,0 +1,527 @@
+#include "sound/dac.h"
+
+#include "speaker.h"
+
+#include "mpu1.lh"
+#include "m_bappl2.lh"
+#include "m_bapple.lh"
+#include "m_gndgit.lh"
+#include "m_lndg.lh"
+#include "m_mtchit.lh"
+#include "m_mtchup.lh"
+
+class mpu1_state : public driver_device
+{
+public:
+	mpu1_state(const machine_config &mconfig, device_type type, const char *tag) :
+		driver_device(mconfig, type, tag),
+		m_maincpu(*this, "maincpu"),
+		m_pia1(*this, "pia1"),
+		m_pia2(*this, "pia2"),
+		m_lamps(*this, "lamp%u", 0U),
+		m_dac(*this, "dac"),
+		m_samples(*this, "samples")
+	{ }
+
+	DECLARE_INPUT_CHANGED_MEMBER(coin_input);
+
+	void mpu1(machine_config &config);
+	void mpu1_lg(machine_config &config);
+
+	void init_mpu1();
+	void init_mpu1_lg();
+
+protected:
+	virtual void machine_start() override;
+	virtual void machine_reset() override;
+
+private:
+	template <unsigned Lamp> DECLARE_WRITE_LINE_MEMBER(pia_lamp_w) { m_lamps[Lamp] = state; }
+
+	void pia1_portb_w(uint8_t data);
+	void pia1_portb_lg_w(uint8_t data);
+	uint8_t pia2_porta_r();
+	void pia2_porta_w(uint8_t data);
+	void pia2_portb_w(uint8_t data);
+	void payout_cash_w(bool state);
+	void payout_token_w(bool state);
+	void meter_w(int meter, bool state);
+	void coin_lockout_w(bool state);
+	void reel_w(int reel, bool state);
+
+	void draw_reel(int reel);
+	void postload();
+
+	void mpu1_map(address_map &map);
+
+	uint8_t m_reel_select;
+	bool m_pia2a_select;
+	uint8_t m_reel_state[4];
+	int m_reel_pos[4];
+	emu_timer *m_reel_timer[4];
+	attotime m_reel_speed;
+	int m_startup_time;
+
+	TIMER_DEVICE_CALLBACK_MEMBER(nmi);
+	template <unsigned Reel> TIMER_CALLBACK_MEMBER(reel_move);
+
+	enum
+	{
+		REEL_STOPPED = 0,
+		REEL_SPINNING,
+		REEL_STOPPING
+	};
+
+	// devices
+	required_device<cpu_device> m_maincpu;
+	required_device<pia6821_device> m_pia1;
+	required_device<pia6821_device> m_pia2;
+	output_finder<13> m_lamps;
+	required_device<dac_1bit_device> m_dac;
+	required_device<fruit_samples_device> m_samples;
+};
+
+void mpu1_state::mpu1_map(address_map &map)
+{
+	map.global_mask(0x3fff);
+	map(0x0000, 0x007f).ram();
+	map(0x0600, 0x0fff).rom();
+	map(0x1000, 0x17ff).rom().region("maskrom", 0).mirror(0x2800);
+	map(0x2000, 0x2003).rw(m_pia1, FUNC(pia6821_device::read), FUNC(pia6821_device::write));
+	map(0x2004, 0x2007).rw(m_pia2, FUNC(pia6821_device::read), FUNC(pia6821_device::write));
+}
+
+void mpu1_state::pia1_portb_w(uint8_t data)
+{
+	if(BIT(data, 7) == 0)
+	{
+		reel_w(0, BIT(data, 0));
+		reel_w(1, BIT(data, 1));
+		reel_w(2, BIT(data, 2));
+		reel_w(3, BIT(data, 3));
+		coin_lockout_w(BIT(data, 4));
+		m_lamps[11] = BIT(data, 5);
+		meter_w(0, BIT(data, 6));
+	}
+	else
+	{
+		m_lamps[12] = BIT(data, 0);
+		payout_cash_w(BIT(data, 1));
+		payout_token_w(BIT(data, 2));
+		meter_w(1, BIT(data, 3));
+		meter_w(2, BIT(data, 4));
+		meter_w(3, BIT(data, 5));
+		meter_w(4, BIT(data, 6));
+	}
+}
+
+void mpu1_state::pia1_portb_lg_w(uint8_t data)
+{
+	if(data == 0) for(int i = 0; i < 5; i++) meter_w(i, 0);
```

Comment:
Usually games just pulse the meters without any issues, but m_bappl2 is a bit weird. It sets a meter's bit high, then to set it back low it just writes a 0 to the port, unlike the Barcrest code which explicitly keeps bit 7. That check sets all the meters back to low if the value is 0, because anything with bit 7 set won't be reached otherwise. Feels a bit hacky, but the meters don't count properly without this. I realized I need to do the same thing for the payouts as well, and I added proper checking to them, otherwise the samples sound weird.